File: app/src/main/java/ninja/dudley/yamr/model/Page.java
package ninja.dudley.yamr.model;

import android.content.ContentValues;
import android.database.Cursor;
import android.net.Uri;

import ninja.dudley.yamr.db.DBHelper;
import ninja.dudley.yamr.db.util.Column;
import ninja.dudley.yamr.db.util.ForeignKey;
import ninja.dudley.yamr.db.util.Table;

/**
 * Created by mdudley on 5/19/15.
 */
@Table(Page.tableName)
public class Page extends MangaElement
{
    public static final String tableName = "page";

    public static final String chapterIdCol = Chapter.tableName + DBHelper.ID;
    @ForeignKey(value = Chapter.class, name = chapterIdCol)
    public final int chapterId;

    public static final String numberCol = "number";
    @Column(name = numberCol, type = Column.Type.Real)
    public float number;

    public static final String imageUrlCol = "image_url";
    @Column(name = imageUrlCol)
    public String imageUrl;

    public static final String imagePathCol = "image_path";
    @Column(name = imagePathCol)
    public String imagePath;


    // Uri Handling --------------------------------------------------------------------------------
    public static Uri baseUri()
    {
        return Uri.parse("content://" + DBHelper.AUTHORITY + "/page");
    }

    public static Uri uri(int id)
    {
        return baseUri().buildUpon().appendPath(Integer.toString(id)).build();
    }

    public Uri uri()
    {
        return uri(id);
    }

    public static Uri heritage(int id)
    {
        return uri(id).buildUpon().appendPath("heritage").build();
    }

    public Uri heritage()
    {
        return heritage(id);
    }

    // Construction / Persistence ------------------------------------------------------------------
    public Page() { chapterId = -1; }
    public Page(int chapterId) { this.chapterId = chapterId; }
    public Page(Cursor c)
    {
        super(c);
        chapterId = getInt(c, chapterIdCol);
        number = getFloat(c, numberCol);
        imageUrl = getString(c, imageUrlCol);
        imagePath = getString(c, imagePathCol);
        c.close();
    }

    @Override
    public ContentValues getContentValues()
    {
        ContentValues values = super.getContentValues();
        values.put(chapterIdCol, chapterId);
        values.put(numberCol, number);
        values.put(imageUrlCol, imageUrl);
        values.put(imagePathCol, imagePath);
        return values;
    }
}

File: app/src/main/kotlin/ninja/dudley/yamr/model/Page.kt
package ninja.dudley.yamr.model

import android.content.ContentValues
import android.database.Cursor
import android.net.Uri

import ninja.dudley.yamr.db.DBHelper
import ninja.dudley.yamr.db.util.Column
import ninja.dudley.yamr.db.util.ForeignKey
import ninja.dudley.yamr.db.util.Table

/**
* Created by mdudley on 5/19/15. Yup.
*/
@Table(Page.tableName)
class Page : MangaElement
{
    @ForeignKey(value = Chapter::class, name = chapterIdCol)
    val chapterId: Int
    @Column(name = numberCol, type = Column.Type.Real)
    val number: Float
    @Column(name = imageUrlCol)
    var imageUrl: String? = null
    @Column(name = imagePathCol)
    var imagePath: String? = null

    // Construction / Persistence ------------------------------------------------------------------
    constructor(chapterId: Int, url: String, number: Float) :
        super(url, MangaElement.UriType.Page)
    {
        this.chapterId = chapterId
        this.number = number
    }

    constructor(c: Cursor, close: Boolean = true) : super(c)
    {
        if (type != MangaElement.UriType.Page)
        {
            throw IllegalArgumentException("Attempted to make a page from a $type")
        }
        chapterId = getInt(c, chapterIdCol)
        number = getFloat(c, numberCol)
        imageUrl = getString(c, imageUrlCol)
        imagePath = getString(c, imagePathCol)
        if (close)
        {
            c.close()
        }
    }

    override fun getContentValues(): ContentValues
    {
        val values = super.getContentValues()
        values.put(chapterIdCol, chapterId)
        values.put(numberCol, number)
        values.put(imageUrlCol, imageUrl)
        values.put(imagePathCol, imagePath)
        return values
    }

    // Uri Handling --------------------------------------------------------------------------------
    fun uri(): Uri
    {
        return uri(id)
    }

    fun heritage(): Uri
    {
        return heritage(id)
    }

    companion object
    {
        const val tableName: String = "page"
        const val chapterIdCol: String = Chapter.tableName + DBHelper.ID
        const val numberCol: String = "number"
        const val imageUrlCol: String = "image_url"
        const val imagePathCol: String = "image_path"


        // Uri Handling ----------------------------------------------------------------------------
        fun baseUri(): Uri
        {
            return Uri.parse("content://" + DBHelper.AUTHORITY + "/page")
        }

        fun uri(id: Int): Uri
        {
            return baseUri().buildUpon().appendPath(Integer.toString(id)).build()
        }

        fun heritage(id: Int): Uri
        {
            return uri(id).buildUpon().appendPath("heritage").build()
        }
    }
}
